### ISSUE TITLE
Add compiler hints to assertions

### ISSUE THREAD

**Issue by chrispellett** [thomhurst/TUnit#1296]
Hello!
First, I just want to express how impressed I am with this project. Both in terms of the improvements it makes on the existing frameworks like xunit, but also on the documentation, extensibility, etc. very impressive!

I wanted to see if it was possible to add some of the compiler culpability hints to the "results" of assertions?

For example, in xunit, I can do:
`Assert.NotNull(x)`

Which has a [NotNull] hint on the parameter. So, after this line, we know the `x` is not null.

This would really help with porting from xunit without adding !s everywhere 

Thanks again!

**Comment by thomhurst** [comment]
Yeah I really want to do this, but I'm fighting the type system and generics.

Trying to convert a `T?` to `T`, especially when wrapped in assertion types `IValueSource<T>`, is making it extremely difficult.

Unfortunately because the `IsNotNull()` method is just an extension, and doesn't have access to the raw value, I can't just annotate it with `[NotNull]`. If you've got any ideas, I'm open to them.

I could possible add to the assertion class `IsNotNull`, which could be used instead of `That` - But I don't really want to because I want to keep things simple.

e.g. `Assert.IsNotNull(someValue).And.IsNotEmpty()`

@vbreuss any ideas from you?

**Comment by vbreuss** [comment]
Unfortunately there is no way to correctly add the `[NotNull]` with this approach, as you would have to specify it within the `Assert.That` method, but you don't know anything yet.
I am working on a [similar project to this assertion library](https://github.com/aweXpect/aweXpect) and I solved it there by allowing you to overwrite your value and infer the nullability in the awaited result:
![image](https://github.com/user-attachments/assets/c54c25a2-29b9-4e30-aaac-94a43b9ec068)

Note that in the first test, the nullability warning appears, but in the second test not (even though the object is indeed `null`). This is correct, because I guarantee, that the await throws an exception in this case.

**Comment by thomhurst** [comment]
Yeah I was able to cast it in the await too, but unfortunately doesn't allow chaining And/Or methods with the new nullability without reassigning the variable.

It's better than nothing I suppose though, and would probably solve most usages of it

**Comment by viceroypenguin** [comment]
Honestly, this whole thread is why I prefer the Assert.Xxx() model used by xUnit. I rarely (if ever) find myself needing the flexibility stuff, and the direct assertions make it easier to reason nullability and such. But that's just my preference.

**Comment by vbreuss** [comment]
> Yeah I was able to cast it in the await too, but unfortunately doesn't allow chaining And/Or methods with the new nullability without reassigning the variable.

I don't understand what you mean. I do no casting, it's just that the compiler knows, if the return value can be null or not after the await...

**Comment by afscrome** [comment]
How would this work with `Assert.Multiple()`? In the below case, `foo` could be null in the second await.

```cs
        Foo foo = null;
        using (Assert.Multiple())
        {
            await Assert.That(foo).IsNotNull();
            await Assert.That(foo.Bar).IsNotNull();
        }

    private class Foo
    {
        public string Bar { get; set; }
    }
```

**Comment by thomhurst** [comment]
> How would this work with `Assert.Multiple()`? In the below case, `foo` could be null in the second await.
> 
> ```cs
>         Foo foo = null;
>         using (Assert.Multiple())
>         {
>             await Assert.That(foo).IsNotNull();
>             await Assert.That(foo.Bar).IsNotNull();
>         }
> 
>     private class Foo
>     {
>         public string Bar { get; set; }
>     }
> ```

Similarly if chaining conditions via the `.Or` property
